src/core/server/plugins/plugins_discovery.ts

```diff
@@ -0,0 +1,230 @@
+          mergeMap(subDirs => subDirs.map(subDir => resolve(dir, subDir))),
+          mergeMap(path =>
+            fsStat$(path).pipe(
+              // Filter out non-directory entries from target directories, it's expected that
+              // these directories may contain files (e.g. `README.md` or `package.json`).
+              // We shouldn't silently ignore the entries we couldn't get stat for though.
+              mergeMap(pathStat => (pathStat.isDirectory() ? [path] : [])),
+              catchError(err => [
+                wrapError(PluginDiscoveryErrorType.InvalidPluginDirectory, path, err),
+              ])
+            )
+          ),
+          catchError(err => [wrapError(PluginDiscoveryErrorType.InvalidScanDirectory, dir, err)])
+        );
+      })
+    );
+  }
+
+  /**
+   * Iterates over every entry in `paths` and returns a stream of all paths that
+   * are directories. If path is not a directory or it's impossible to get stat
+   * for this path then error is added into the stream instead.
+   * @param paths List of paths to process.
+   */
+  private processPaths$(paths: string[]) {
+    return from(paths).pipe(
+      mergeMap(path => {
+        this.log.debug(`Including "${path}" into the plugin path list.`);
+
+        return fsStat$(path).pipe(
+          // Since every path is specifically provided we should treat non-directory
+          // entries as mistakes we should report of.
+          mergeMap(pathStat => {
+            return pathStat.isDirectory()
+              ? [path]
+              : throwError(new Error(`${path} is not a directory.`));
+          }),
+          catchError(err => [wrapError(PluginDiscoveryErrorType.InvalidPluginDirectory, path, err)])
+        );
+      })
+    );
+  }
+
+  /**
+   * Tries to load and parse the plugin manifest file located at the provided
+   * plugin directory path and throws if it fails to do so or plugin manifest
+   * isn't valid.
+   * @param path Path to the plugin directory where manifest should be loaded from.
+   */
+  private createPlugin$(path: string) {
+    const manifestPath = resolve(path, MANIFEST_FILE_NAME);
+    return fsReadFile$(manifestPath).pipe(
+      mergeMap(manifestContent => {
+        try {
+          const plugin = { path, manifest: parseManifest(manifestContent) };
+
+          this.log.debug(`Successfully discovered plugin "${plugin.manifest.id}" at "${path}"`);
+
+          return [{ plugin }];
+        } catch (err) {
+          return [wrapError(PluginDiscoveryErrorType.InvalidManifest, manifestPath, err)];
+        }
+      }),
+      catchError(err => [wrapError(PluginDiscoveryErrorType.MissingManifest, manifestPath, err)])
+    );
+  }
+}
+
+/**
+ * Parses raw buffer content into plugin manifest with the preliminary checks.
+ * @param rawManifest Buffer containing plugin manifest JSON.
+ */
+function parseManifest(rawManifest: Buffer): PluginManifest {
+  const manifest = JSON.parse(rawManifest.toString());
+  if (
```

Comment:
__note:__ I made discovery a bit dumb so that it doesn't check plugin vs kibana versions match and defer it to the `PluginService`, it may change in the next PR, but what are your thoughts?

Reply:
Okay, noted!